### PR TITLE
Close #68 - prepare_release tool (version bump + CHANGELOG roll-up + release PR)

### LINKED ISSUE #68: feat: prepare_release tool — version bump + CHANGELOG roll-up + release PR
## Goal

Package the deterministic, error-prone release-prep steps we currently do by hand (during 0.1.6 the `package-lock.json` sync was missed and the CHANGELOG/compare links were hand-rolled). A `prepare_release` tool automates the mechanical parts and stops before anything irreversible.

## Scope: prepare, not publish

The tool **prepares** a release; it does **not** tag or publish. Tagging + npm publish stay with the existing CI workflow (`publish.yml` fires on a `v*.*.*` tag push, which a human triggers). This keeps the irreversible step human-gated.

## Behavior

Given a target version, the tool:
1. Bumps the version in **`package.json` and `package-lock.json`** (both self-version fields).
2. Rolls the CHANGELOG: rename `## [Unreleased]` → `## [X.Y.Z] - <today>`, insert a fresh empty `## [Unreleased]` above it, and update/add the compare links at the bottom (`[Unreleased]` → compare `vX.Y.Z...HEAD`, add `[X.Y.Z]` → `vPREV...vX.Y.Z`).
3. Commits the changes on a release branch and opens a PR for review.
4. Returns a summary (new version, changelog diff, PR URL) and a reminder that tagging `vX.Y.Z` after merge triggers the CI publish.

## Inputs

- `version` (explicit, e.g. `0.1.7`) **or** `bump` (`patch` | `minor` | `major`) — exactly one. **Decision: explicit input is authoritative; the user stays in control of semver.**
- If neither is given, **infer a suggested default** from the `[Unreleased]` entries (has `### Added` → minor; only `### Fixed`/`### Changed` → patch) and surface it for confirmation rather than acting silently.
- `confirmed` (two-step): first call previews the version + changelog transformation; re-call with `confirmed: true` to apply.

## Does NOT do

- No git tag, no `npm publish` (CI handles those on the human-pushed tag).
- No merge — opens a PR for review.

## Implementation notes

- CHANGELOG transformation logic belongs near the existing `docs.ts` helpers (we already parse `[Unreleased]` sections there).
- Bump both `package.json` and `package-lock.json` self-version fields (the bit missed manually in 0.1.6).
- Branch/PR target: create a release branch off `OKFFS_BASE_BRANCH` and PR back to it (matches the develop-based flow); the subsequent develop→main + tag stays manual. Confirm exact target during implementation.

## Notes

Nicely tool-shaped (deterministic) — unlike the review/guidance workflows that needed prompts. Good first user of itself: once built, use `prepare_release` to cut the 0.1.7 release that ships it.

## Labels
`enhancement`, `okffs`

**Branch:** `68-okffs-feat-preparerelease-tool-version-bump`

### PR DESCRIPTION
**Pull request by 2b9sa2owa**
Closes #68

Adds a `prepare_release` tool that automates the deterministic, error-prone release-prep steps (the bits hand-done — and partly missed — during 0.1.6).

**What it does**
- Bumps the version in `package.json` **and** `package-lock.json` (targeted version-field edits, no full reformat).
- Rolls the CHANGELOG: `[Unreleased]` → `## [X.Y.Z] - <date>`, leaves a fresh empty `[Unreleased]`, and updates/adds the compare links (helpers `getUnreleasedSection` + `rollChangelogForRelease` in `docs.ts`).
- Commits on a `release/X.Y.Z` branch off `OKFFS_BASE_BRANCH` and opens a PR.

**Inputs**
- Explicit `version` (e.g. `0.2.0`) or a `bump` level; if neither, the level is inferred from the `[Unreleased]` entries (`### Added` present → minor, else patch) and surfaced for confirmation.
- Two-step: first call previews, re-call with `confirmed: true` to apply.

**Deliberately does NOT** tag or publish — tagging `vX.Y.Z` after merge triggers the CI npm publish, keeping the irreversible step a manual decision.

**Verified**: changelog roll transformation (fresh `[Unreleased]`, dated section, correct compare links), version inference (suggested `0.2.0` for the current unreleased set, which is correct — it's a feature release), and that the lockfile has exactly two self-version fields so the targeted replace is safe.

Docs: README (Tools table + rewritten Publishing section) and CLAUDE.md.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/tools/prepare_release.ts [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+  let out = content;
+  let from = 0;
+  for (let i = 0; i < count; i++) {
+    const at = out.indexOf(search, from);
+    if (at === -1) break;
+    out = out.slice(0, at) + replace + out.slice(at + search.length);
+    from = at + replace.length;
+  }
+  return out;
+}
```
[the remark this answers is not in the document]
Comment (2b9sa2owa) [reply src/tools/prepare_release.ts]: Fixed in 9d28b66. `replaceFirst` is now `replaceExactly`, which counts occurrences and **throws if fewer than required** are found. Also, all new file contents (package.json, package-lock.json, CHANGELOG) are now computed up front before any write — so a validation failure aborts cleanly with nothing written and nothing committed, instead of a partial bump.

File: src/tools/prepare_release.ts [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+    content: [{
+      type: "text" as const,
+      text:
+        `Prepared release ${targetVersion} (from ${currentVersion}).\n` +
+        `Branch: ${branch}\nPR: ${pr.html_url}\n\n` +
+        `Next: review & merge the PR, then tag \`v${targetVersion}\` and push it — CI publishes to npm. ` +
+        `prepare_release does not tag or publish.`,
+    }],
+  };
+}
```
[the remark this answers is not in the document]
Comment (2b9sa2owa) [reply src/tools/prepare_release.ts]: Fixed in 9d28b66. `createPullRequest` is now wrapped in try/catch. If PR creation fails after the branch is pushed, the tool returns the branch name and manual next steps (open a PR from the branch into the base, then tag `vX.Y.Z` after merge) instead of throwing.

File: src/docs.ts [hunk 3]
```diff
@@ -61,6 +61,52 @@ function buildChangelogEntry(ctx: DocsContext): string {
...
+  result = result.replace(
+    /\[Unreleased\]:\s*\S+/,
+    `[Unreleased]: ${repoUrl}/compare/v${version}...HEAD`
+  );
+  result = result.replace(
+    /(\[Unreleased\]:[^\n]*\n)/,
+    `$1[${version}]: ${repoUrl}/compare/v${prevVersion}...v${version}\n`
+  );
+  return result;
+}
```
[the remark this answers is not in the document]
Comment (2b9sa2owa) [reply src/docs.ts]: Fixed in 9d28b66. `rollChangelogForRelease` now: (1) **appends** a compare-links block when none exists (new/manually-edited changelogs), and (2) **skips** inserting the `[version]` ref if it's already present — so re-running for the same version no longer creates a duplicate. Verified both with isolated tests (re-run yields a single `[0.2.0]` ref; a link-less changelog gets both refs appended).

**Comment by 2b9sa2owa** [comment]
## Review fixes applied (9d28b66)

| Comment | Fix |
|---------|-----|
| `replaceFirst` silently stops early → partial version bump | Now `replaceExactly`: fails fast if fewer than the required occurrences exist. All new file contents are computed before any write, so a failure aborts with nothing written/committed. |
| `createPullRequest` not caught → unhelpful throw after push | Wrapped in try/catch; returns the branch name + manual next steps (open PR, then tag) on failure. |
| `rollChangelogForRelease` compare links: missing links not added; duplicate `[version]` ref on re-run | Appends a links block when none exists; skips the `[version]` ref if already present. Verified with isolated tests. |

No change to the core release flow — robustness/edge-case hardening. Build passes.
